xds/src/main/java/io/grpc/xds/LookasideLb.java

```diff
@@ -82,60 +101,44 @@
+    this.switchingLoadBalancer = new GracefulSwitchLoadBalancer(lookasideLbHelper);
     this.localityStoreFactory = localityStoreFactory;
     this.loadReportClientFactory = loadReportClientFactory;
-  }
-
-  @Override
-  protected LoadBalancer delegate() {
-    return lookasideChannelLb;
+    this.bootstrapper = bootstrapper;
   }
 
   @Override
   public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
     // In the future, xdsConfig can be gotten directly by
     // resolvedAddresses.getLoadBalancingPolicyConfig()
     Attributes attributes = resolvedAddresses.getAttributes();
+    final ObjectPool<XdsClient> xdsClientRefFromResolver =
+        resolvedAddresses.getAttributes().get(XdsAttributes.XDS_CLIENT_REF);
+    if (xdsClientRefFromResolver != null) {
+      cachedXdsClientRef = xdsClientRefFromResolver;
+    }
     Map<String, ?> newRawLbConfig = checkNotNull(
         attributes.get(ATTR_LOAD_BALANCING_CONFIG), "ATTR_LOAD_BALANCING_CONFIG not available");
     ConfigOrError cfg =
         XdsLoadBalancerProvider.parseLoadBalancingConfigPolicy(newRawLbConfig, lbRegistry);
     if (cfg.getError() != null) {
       throw cfg.getError().asRuntimeException();
     }
-    XdsConfig xdsConfig = (XdsConfig) cfg.getConfig();
-
-    final String newBalancerName = xdsConfig.balancerName;
-
-    // The is to handle the legacy usecase that requires balancerName from xds config.
-    if (!newBalancerName.equals(balancerName)) {
-      balancerName = newBalancerName; // cache the name and check next time for optimization
-      Node nodeFromResolvedAddresses = resolvedAddresses.getAttributes().get(XDS_NODE);
-      final Node node;
-      if (nodeFromResolvedAddresses == null) {
-        node = Node.newBuilder()
-            .setMetadata(Struct.newBuilder()
-                .putFields(
-                    "endpoints_required",
-                    Value.newBuilder().setBoolValue(true).build()))
-            .build();
-      } else {
-        node = nodeFromResolvedAddresses;
-      }
-      List<ChannelCreds> channelCredsListFromResolvedAddresses =
-          resolvedAddresses.getAttributes().get(XDS_CHANNEL_CREDS_LIST);
-      final List<ChannelCreds> channelCredsList;
-      if (channelCredsListFromResolvedAddresses == null) {
-        channelCredsList = Collections.emptyList();
-      } else {
-        channelCredsList = channelCredsListFromResolvedAddresses;
-      }
+    final XdsConfig newXdsConfig = (XdsConfig) cfg.getConfig();
 
-      LoadBalancerProvider childBalancerProvider = new LoadBalancerProvider() {
+    // If XdsConfig is changed, do a graceful switch.
+    if (!newXdsConfig.equals(xdsConfig)) {
+      final XdsConfig oldXdsConfig = xdsConfig;
+      final EndpointWatcher oldEndpointWatcher = endpointWatcher;
+      xdsConfig = newXdsConfig;
+
+      // Provides a load balancer with the fixed XdsConfig: newXdsConfig.
+      LoadBalancerProvider fixedXdsConfigBalancer = new LoadBalancerProvider() {
```

Comment:
This local class is too big. Should make it a standalone class and use constructor to create instance.

Reply:
I agree. Also:
- the variable should be named `fixedXdsConfigBalancerProvider` 
- there are 2 anonymous classes and the inner `new LoadBalancer()` is really the big one. How about making it a non-anonynous inner class?
- similar pattern to the CDSLoadBalancer PR I just reviewed. At some point if we can refactor to eliminate code duplication that will be good

Reply:
Made both non-anonymous. I found only making `LoadBalancer` non-anonymous still introduces too many levels of indentation.